File: agentscope-service/aistio/sdk/python/aistio/collaboration.py
# Copyright 2024-2026 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Backend-independent Issue/Comment/AgentTask client for Agent runtimes."""
from __future__ import annotations

import json
import mimetypes
import uuid
from dataclasses import dataclass
from typing import Any
from urllib.error import HTTPError
from urllib.parse import quote, urlencode
from urllib.request import Request, urlopen


class CollaborationError(RuntimeError):
    def __init__(self, operation: str, status: int, body: str) -> None:
        super().__init__(f"{operation} failed: HTTP {status}: {body}")
        self.operation, self.status, self.body = operation, status, body


@dataclass
class CollaborationClient:
    """Same collaboration contract for Managed, External, and Hosted Agents."""
    base_url: str
    internal_token: str = ""
    timeout: float = 10.0

    def issue(self, issue_id: str, task_token: str) -> dict[str, Any]:
        return self._send("GET", f"/api/v1/issues/{quote(issue_id)}", operation="issue.get", headers={"X-Agent-Task-Token": task_token})

    def comments(self, issue_id: str, task_token: str, **query: Any) -> dict[str, Any]:
        suffix = urlencode({k: v for k, v in query.items() if v is not None})
        return self._send("GET", f"/api/v1/issues/{quote(issue_id)}/comments" + (f"?{suffix}" if suffix else ""), operation="issue.comment.list", headers={"X-Agent-Task-Token": task_token})

    def add_comment(self, issue_id: str, task_token: str, content: str, *, parent_id: str | None = None, mentions: list[dict[str, str]] | None = None, comment_type: str = "comment") -> dict[str, Any]:
        return self._send("POST", f"/api/v1/issues/{quote(issue_id)}/comments", {"content": content, "parentId": parent_id, "mentions": mentions or [], "type": comment_type}, operation="issue.comment.add", headers={"X-Agent-Task-Token": task_token})

    def create_child_from_task(self, task_id: str, task_token: str, body: dict[str, Any]) -> dict[str, Any]:
        return self._task_send("POST", task_id, "children", task_token, body, "issue.child.create")

    def team(self, team_id: str, task_token: str) -> dict[str, Any]:
        return self._send("GET", f"/api/v1/teams/{quote(team_id)}", operation="team.get", headers={"X-Agent-Task-Token": task_token})

    def request_approval(self, task_token: str, body: dict[str, Any]) -> dict[str, Any]:
        return self._send("POST", "/api/v1/approvals", body, operation="approval.request", headers={"X-Agent-Task-Token": task_token})

    def task(self, task_id: str, task_token: str) -> dict[str, Any]:
        return self._send("GET", f"/api/v1/agent-tasks/{quote(task_id)}", operation="task.get", headers={"X-Agent-Task-Token": task_token})

    def task_context(self, task_id: str, task_token: str) -> dict[str, Any]:
        return self._task_send("GET", task_id, "context", task_token, None, "task.context")

    def ack(self, task_id: str, task_token: str, input_ids: list[str]) -> dict[str, Any]:
        return self._task_send("POST", task_id, "ack", task_token, {"inputIds": input_ids}, "task.ack")

    def start(self, task_id: str, task_token: str, *, expected_version: int = 0) -> dict[str, Any]:
        return self._task_send("POST", task_id, "start", task_token, {"expectedVersion": expected_version}, "task.start")

    def progress(self, task_id: str, task_token: str, content: str, *, mentions: list[dict[str, str]] | None = None) -> dict[str, Any]:
        return self._task_send("POST", task_id, "progress", task_token, {"content": content, "mentions": mentions or []}, "task.progress")

    def respond(self, task_id: str, task_token: str, content: str, *, parent_id: str | None = None, mentions: list[dict[str, str]] | None = None) -> dict[str, Any]:
        return self._task_send("POST", task_id, "respond", task_token, {"content": content, "parentId": parent_id, "mentions": mentions or [], "type": "result"}, "task.respond")

    def complete(self, task_id: str, task_token: str, *, summary: str = "", result: Any = None, processed_input_ids: list[str] | None = None, deferred_input_ids: list[str] | None = None, expected_version: int = 0) -> dict[str, Any]:
        body = {"summary": summary, "result": result or {}, "processedInputIds": processed_input_ids or [], "deferredInputIds": deferred_input_ids or [], "expectedVersion": expected_version}
        return self._task_send("POST", task_id, "complete", task_token, body, "task.complete")

    def fail(self, task_id: str, task_token: str, code: str, message: str, *, expected_version: int = 0) -> dict[str, Any]:
        return self._task_send("POST", task_id, "fail", task_token, {"code": code, "message": message, "expectedVersion": expected_version}, "task.fail")

    def run(self, task_id: str, task_token: str) -> dict[str, Any]:
        return self._task_send("GET", task_id, "run", task_token, None, "run.get")

    def run_graph(self, task_id: str, task_token: str) -> dict[str, Any]:
        return self._task_send("GET", task_id, "run/graph", task_token, None, "run.graph")

    def complete_run_node(self, task_id: str, task_token: str, output: Any) -> dict[str, Any]:
        return self._task_send("POST", task_id, "run/node/complete", task_token, {"output": output}, "run.node.complete")

    def fail_run_node(self, task_id: str, task_token: str, code: str, message: str) -> dict[str, Any]:
        return self._task_send("POST", task_id, "run/node/fail", task_token, {"code": code, "message": message}, "run.node.fail")

    def replan_run(self, task_id: str, task_token: str, node: dict[str, Any]) -> dict[str, Any]:
        return self._task_send("POST", task_id, "run/replan", task_token, node, "run.replan")

    def signal_run(self, task_id: str, task_token: str, name: str, idempotency_key: str, payload: Any = None) -> dict[str, Any]:
        return self._task_send("POST", task_id, f"run/signals/{quote(name)}", task_token, {"idempotencyKey": idempotency_key, "payload": payload}, "run.signal")

    def run_artifacts(self, task_id: str, task_token: str) -> dict[str, Any]:
        return self._task_send("GET", task_id, "run/artifacts", task_token, None, "run.artifacts")

    def upload_artifact(self, task_id: str, task_token: str, filename: str, content: bytes, *, content_type: str | None = None, target_type: str = "agent-task", target_ref: str | None = None) -> dict[str, Any]:
        boundary = "aistio-" + uuid.uuid4().hex
        media_type = content_type or mimetypes.guess_type(filename)[0] or "application/octet-stream"
        fields = {"sourceTaskId": task_id, "targetType": target_type, "targetRef": target_ref or task_id}
        parts: list[bytes] = []
        for key, value in fields.items():
            parts.append(f"--{boundary}\r\nContent-Disposition: form-data; name=\"{key}\"\r\n\r\n{value}\r\n".encode())
        safe_name = filename.replace('"', "_").replace("\r", "_").replace("\n", "_")
        parts.append(f"--{boundary}\r\nContent-Disposition: form-data; name=\"file\"; filename=\"{safe_name}\"\r\nContent-Type: {media_type}\r\n\r\n".encode() + content + b"\r\n")
        parts.append(f"--{boundary}--\r\n".encode())
        return self._send_bytes("POST", "/api/v1/artifacts/uploads", b"".join(parts), operation="artifact.upload", headers={"Content-Type": f"multipart/form-data; boundary={boundary}", "X-Agent-Task-Token": task_token})

    def download_artifact(self, artifact_id: str, task_id: str, task_token: str) -> bytes:
        query = urlencode({"taskId": task_id})
        request = Request(self.base_url.rstrip("/") + f"/api/v1/artifacts/{quote(artifact_id)}/download?{query}", headers=self._headers({"X-Agent-Task-Token": task_token}), method="POST")
        try:
            with urlopen(request, timeout=self.timeout) as response:
                return response.read()
        except HTTPError as exc:
            raise CollaborationError("artifact.download", exc.code, exc.read().decode(errors="replace")) from exc

    def _task_send(self, method: str, task_id: str, action: str, token: str, body: Any, operation: str) -> dict[str, Any]:
        return self._send(method, f"/api/v1/agent-tasks/{quote(task_id)}/{action}", body, operation=operation, headers={"X-Agent-Task-Token": token})

    def _send(self, method: str, path: str, body: Any = None, *, operation: str, headers: dict[str, str] | None = None) -> dict[str, Any]:
        request_headers = self._headers(headers)
        data = None
        if body is not None:
            request_headers["Content-Type"] = "application/json"
            data = json.dumps(body).encode()
        request = Request(self.base_url.rstrip("/") + path, data=data, headers=request_headers, method=method)
        try:
            with urlopen(request, timeout=self.timeout) as response:
                raw = response.read()
                return json.loads(raw) if raw else {}
        except HTTPError as exc:
            raise CollaborationError(operation, exc.code, exc.read().decode(errors="replace")) from exc

    def _send_bytes(self, method: str, path: str, body: bytes, *, operation: str, headers: dict[str, str] | None = None) -> dict[str, Any]:
        request = Request(self.base_url.rstrip("/") + path, data=body, headers=self._headers(headers), method=method)
        try:
            with urlopen(request, timeout=self.timeout) as response:
                raw = response.read()
                return json.loads(raw) if raw else {}
        except HTTPError as exc:
            raise CollaborationError(operation, exc.code, exc.read().decode(errors="replace")) from exc

    def _headers(self, headers: dict[str, str] | None = None) -> dict[str, str]:
        result = {"Accept": "application/json", **(headers or {})}
        if self.internal_token:
            result["X-Builder-Internal-Token"] = self.internal_token
        return result

File: agentscope-service/frontend/src/api/collaboration.ts
/*
 * Copyright 2024-2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { api, apiFetch, apiResponse } from "@/lib/apiClient";

export interface Actor {
  type: "human" | "agent" | "system" | "automation";
  ref?: string;
}
export interface Issue {
  access?: import("./permissions").IssueAccess;
  id: string;
  identifier?: string;
  tenant: string;
  namespace: string;
  title: string;
  description?: string;
  status: string;
  priority: string;
  kind: 'user_work' | 'endpoint_job' | 'automation_job' | string;
  visibility: 'work_hub' | 'operational' | string;
  completionPolicy: 'review' | 'automatic' | 'external' | string;
  assigneeType?: string;
  assigneeRef?: string;
  executionTargetType?: "agent" | "team" | "orchestration_revision" | string;
  executionTargetRef?: string;
  creator: Actor;
  parentIssueId?: string;
  acceptanceCriteria?: unknown;
  contextRefs?: unknown;
  sourceType?: string;
  sourceRef?: string;
  dueAt?: string;
  version: number;
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
  archivedAt?: string;
}
export interface IssueSummary {
  issueId: string;
  title: string;
  status: string;
  commentCount: number;
  unresolvedThreads: number;
  activeTasks: number;
  terminalTasks: number;
  childCount: number;
  latestResult?: string;
  updatedAt: string;
}
export interface Comment {
  id: string;
  issueId: string;
  parentId?: string;
  threadRootId: string;
  author: Actor;
  content: string;
  type: string;
  sourceTaskId?: string;
  version: number;
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
  deletedAt?: string;
  mentions?: Array<{ targetType: string; targetRef: string }>;
  routes?: Array<{
    targetType: string;
    targetRef: string;
    outcome: string;
    reasonCode?: string;
    taskId?: string;
  }>;
  externalSync?: {
    syncState: string;
    lastError?: string;
    attempts: number;
    externalId?: string;
  };
}
export interface IssueActivity {
  id: string;
  issueId?: string;
  actor: Actor;
  action: string;
  objectType: string;
  objectRef: string;
  causationId?: string;
  correlationId?: string;
  details?: Record<string, unknown>;
  createdAt: string;
}
export interface IssueSubscriber {
  issueId: string;
  subscriberType: string;
  subscriberRef: string;
  createdAt: string;
}
export interface Artifact {
  id: string;
  filename: string;
  contentType: string;
  sizeBytes: number;
  checksum: string;
  uploader: Actor;
  createdAt: string;
  expiresAt?: string;
}
export interface TaskInput {
  id: string;
  commentId: string;
  commentVersion: number;
  sequence: number;
  state: string;
  attempts: number;
}
export interface AgentTask {
  id: string;
  tenant: string;
  namespace: string;
  issueId: string;
  orchestrationRunId: string;
  runNodeId: string;
  currentAttemptId?: string;
  agentId: string;
  status: string;
  priority: number;
  triggerType: string;
  retryOfTaskId?: string;
  rerunOfTaskId?: string;
  teamId?: string;
  teamRole?: string;
  leaderTask?: boolean;
  originator: Actor;
  runtimeBinding?: unknown;
  sessionId?: string;
  result?: unknown;
  errorCode?: string;
  errorMessage?: string;
  version: number;
  createdAt: string;
  dispatchedAt?: string;
  startedAt?: string;
  completedAt?: string;
  inputs?: TaskInput[];
}
export interface RuntimeBinding {
  agentId: string;
  bindingId: string;
  kind: "managed" | "external-application" | "hosted-runtime";
  ownerRef?: string;
  managedDefinitionRef?: string;
  instanceSelector?: Record<string, string>;
  runtimeProfileId?: string;
  runtimePoolId?: string;
}
export interface RuntimeBindingCandidate {
  binding: RuntimeBinding;
  requiredCapabilities?: Record<string, unknown>;
  securityConstraints?: Record<string, unknown>;
}
export interface RuntimeBindingPolicy {
  candidates: RuntimeBindingCandidate[];
  selectionMode: "ordered";
  fallbackMode: "disabled" | "fresh";
  retryPolicy?: Record<string, unknown>;
}
export interface TeamMember {
  id: string;
  teamId: string;
  agentId: string;
  role: string;
  instructions?: string;
  capabilityRequirements?: Record<string, unknown>;
  runtimeBindingPolicy?: RuntimeBindingPolicy | null;
}
export interface Team {
  id: string;
  tenant: string;
  namespace: string;
  name: string;
  description?: string;
  instructions?: string;
  status: 'active' | 'disabled' | string;
  leaderAgentId: string;
  policy?: unknown;
  version: number;
  createdAt: string;
  updatedAt: string;
  members?: TeamMember[];
}
export interface TeamOverview {
  teamId: string;
  observedAt: string;
  readiness: 'ready' | 'degraded' | 'unavailable' | string;
  reason: string;
  members: Array<{
    agentId: string;
    role: string;
    leader: boolean;
    lifecycle?: string;
    readiness: { state: string; mode?: string; reason: string; activeBindingId?: string };
  }>;
  runs: { total: number; activeTasks: number };
  endpoints: { total: number; published: number };
}
export interface InboxItem {
  id: string;
  type: string;
  severity: string;
  issueId?: string;
  commentId?: string;
  approvalId?: string;
  actor: Actor;
  title: string;
  body?: string;
  details?: Record<string, unknown>;
  needsAction: boolean;
  readAt?: string;
  resolvedAt?: string;
  read: boolean;
  archived: boolean;
  createdAt: string;
}
export interface Approval {
  id: string;
  targetType: string;
  targetRef: string;
  issueId?: string;
  runId?: string;
  runNodeId?: string;
  requestedBy: Actor;
  approverRef: string;
  status: string;
  reason?: string;
  request?: unknown;
  decision?: unknown;
  version: number;
  createdAt: string;
  updatedAt: string;
}
export type { Automation } from './automations';
export { listAutomations, createAutomation, updateAutomation, triggerAutomation } from './automations';

function query(values: Record<string, string | number | undefined>) {
  const p = new URLSearchParams();
  Object.entries(values).forEach(([k, v]) => {
    if (v !== undefined && v !== "") p.set(k, String(v));
  });
  const s = p.toString();
  return s ? `?${s}` : "";
}
export const listIssues = (
  tenant: string,
  namespace: string,
  status = "",
  search = "",
  archived = false,
  filters: {
    kind?: string;
    visibility?: string;
    includeOperational?: boolean;
  } = {},
) =>
  api.get<{ items: Issue[]; nextCursor?: string }>(
    `/api/v1/issues${query({
      tenant,
      namespace,
      status,
      search,
      archived: archived ? "true" : undefined,
      kind: filters.kind,
      visibility: filters.visibility,
      includeOperational: filters.includeOperational ? "true" : undefined,
    })}`,
  );
export const getIssue = (id: string) =>
  api.get<{ issue: Issue }>(`/api/v1/issues/${encodeURIComponent(id)}`);
export const createIssue = (body: unknown) =>
  api.post<{ issue: Issue; agentTask?: AgentTask }>("/api/v1/issues", body);
export const updateIssue = (id: string, body: unknown) =>
  api.patch<{ issue: Issue }>(`/api/v1/issues/${encodeURIComponent(id)}`, body);
export const transitionIssue = (
  id: string,
  status: string,
  expectedVersion: number,
  reason = "",
) =>
  api.post<{ issue: Issue }>(
    `/api/v1/issues/${encodeURIComponent(id)}/transition`,
    { status, expectedVersion, reason },
  );
export const acceptIssue = (id: string, expectedVersion: number) =>
  api.post<{ issue: Issue }>(
    `/api/v1/issues/${encodeURIComponent(id)}/accept`,
    { expectedVersion },
  );
export const rejectIssue = (
  id: string,
  expectedVersion: number,
  reason: string,
) =>
  api.post<{ issue: Issue }>(
    `/api/v1/issues/${encodeURIComponent(id)}/reject`,
    { expectedVersion, reason },
  );
export const reopenIssue = (id: string, expectedVersion: number, reason = "") =>
  api.post<{ issue: Issue }>(
    `/api/v1/issues/${encodeURIComponent(id)}/reopen`,
    { expectedVersion, reason },
  );
export const archiveIssue = (id: string, expectedVersion: number) =>
  api.post<{ issue: Issue }>(
    `/api/v1/issues/${encodeURIComponent(id)}/archive`,
    { expectedVersion },
  );
export const getIssueSummary = (id: string) =>
  api.get<{ summary: IssueSummary }>(
    `/api/v1/issues/${encodeURIComponent(id)}/summary`,
  );
export const exportIssue = (id: string) =>
  api.get<Record<string, unknown>>(
    `/api/v1/issues/${encodeURIComponent(id)}/export`,
  );
export const listComments = async (issueId: string, focusCommentId?: string) => {
  const items: Comment[] = [];
  let cursor = "";
  do {
    const response = await api.get<{
      items: Comment[];
      nextCursor?: string;
      externalSync?: Record<string, NonNullable<Comment["externalSync"]>>;
    }>(`/api/v1/issues/${encodeURIComponent(issueId)}/comments${query({ cursor: cursor || undefined })}`);
    items.push(...response.items.map(comment => ({ ...comment, externalSync: response.externalSync?.[comment.id] })));
    const next = response.nextCursor || "";
    if (!focusCommentId || items.some(comment => comment.id === focusCommentId) || !next || next === cursor) break;
    cursor = next;
  } while (cursor);
  return { items };
};
export const listIssueActivity = (issueId: string) =>
  api.get<{ items: IssueActivity[] }>(
    `/api/v1/issues/${encodeURIComponent(issueId)}/activity`,
  );
export const listIssueSubscribers = (issueId: string) =>
  api.get<{ items: IssueSubscriber[] }>(
    `/api/v1/issues/${encodeURIComponent(issueId)}/subscribers`,
  );
export const subscribeIssue = (issueId: string) =>
  api.post<{ subscriber: IssueSubscriber }>(
    `/api/v1/issues/${encodeURIComponent(issueId)}/subscribers`,
    { type: "human" },
  );
export const unsubscribeIssue = (
  issueId: string,
  subscriberType: string,
  subscriberRef: string,
) =>
  api.delete<void>(
    `/api/v1/issues/${encodeURIComponent(issueId)}/subscribers/${encodeURIComponent(subscriberType)}/${encodeURIComponent(subscriberRef)}`,
  );
export const listIssueArtifacts = (issueId: string) =>
  api.get<{ items: Artifact[] }>(
    `/api/v1/issues/${encodeURIComponent(issueId)}/artifacts`,
  );
export async function downloadIssueArtifact(artifact: Pick<Artifact, "id" | "filename">): Promise<Blob> {
  const response = await apiResponse(`/api/v1/artifacts/${encodeURIComponent(artifact.id)}/download`, { method: "POST" });
  return response.blob();
}
export const addComment = (
  issueId: string,
  content: string,
  parentId?: string,
  mentions: Array<{ type: string; ref: string }> = [],
  type = "comment",
) =>
  api.post<Comment>(`/api/v1/issues/${encodeURIComponent(issueId)}/comments`, {
    content,
    parentId,
    mentions,
    type,
  });
export const resolveComment = (
  issueId: string,
  commentId: string,
  expectedVersion: number,
  resolved = true,
) =>
  api.post<{ comment: Comment }>(
    `/api/v1/issues/${encodeURIComponent(issueId)}/comments/${encodeURIComponent(commentId)}/resolve`,
    { expectedVersion, resolved },
  );
export const assignIssue = (
  id: string,
  assigneeType: string,
  assigneeRef: string,
  expectedVersion: number,
) =>
  api.post<{ issue: Issue; agentTask?: AgentTask }>(
    `/api/v1/issues/${encodeURIComponent(id)}/assign`,
    { assigneeType, assigneeRef, expectedVersion },
  );
export const createChildIssue = (id: string, body: unknown) =>
  api.post<{ issue: Issue; agentTask?: AgentTask }>(
    `/api/v1/issues/${encodeURIComponent(id)}/children`,
    body,
  );
export const listChildIssues = (
  tenant: string,
  namespace: string,
  parentIssueId: string,
) =>
  api.get<{ items: Issue[] }>(
    `/api/v1/issues${query({ tenant, namespace, parentIssueId })}`,
  );
export const uploadIssueArtifact = async (
  tenant: string,
  namespace: string,
  issueId: string,
  file: File,
) => {
  const form = new FormData();
  form.set("tenant", tenant);
  form.set("namespace", namespace);
  form.set("targetType", "issue");
  form.set("targetRef", issueId);
  form.set("relation", "attachment");
  form.set("file", file);
  return apiFetch<{
    artifact: {
      id: string;
      filename: string;
      sizeBytes: number;
      contentType: string;
    };
  }>("/api/v1/artifacts/uploads", { method: "POST", body: form });
};
export const listTasks = (
  tenant: string,
  namespace: string,
  status = "",
  agentId = "",
) =>
  api.get<{ items: AgentTask[] }>(
    `/api/v1/agent-tasks${query({ tenant, namespace, status, agentId })}`,
  );
export async function listIssueTasks(tenant: string, namespace: string, issueId: string) {
  const items: AgentTask[] = [];
  for (let offset = 0; ; offset += 100) {
    const page = await api.get<{ items: AgentTask[] }>(
      `/api/v1/agent-tasks${query({ tenant, namespace, issueId, limit: 100, offset })}`,
    );
    items.push(...page.items);
    if (page.items.length < 100) return { items };
  }
}
export const listTeamTasks = (tenant: string, namespace: string, teamId: string) =>
  api.get<{ items: AgentTask[] }>(
    `/api/v1/agent-tasks${query({ tenant, namespace, teamId })}`,
  );
export const getTask = (id: string) =>
  api.get<{ task: AgentTask; inputSummaries?: Array<{ inputId: string; commentId: string; version: number; state: string; content?: string }> }>(`/api/v1/agent-tasks/${encodeURIComponent(id)}`);
export const cancelTask = (id: string, expectedVersion: number) =>
  api.post<{ task: AgentTask }>(
    `/api/v1/agent-tasks/${encodeURIComponent(id)}/cancel`,
    { expectedVersion },
  );
export const retryTask = (id: string) =>
  api.post<{ task: AgentTask }>(
    `/api/v1/agent-tasks/${encodeURIComponent(id)}/retry`,
    {},
  );
export const listTeams = (tenant: string, namespace: string) =>
  api.get<{ items: Team[] }>(`/api/v1/teams${query({ tenant, namespace })}`);
export const getTeam = (id: string) =>
  api.get<{ team: Team }>(`/api/v1/teams/${encodeURIComponent(id)}`);
export const getTeamOverview = (id: string) =>
  api.get<{ team: Team; overview: TeamOverview }>(`/api/v1/teams/${encodeURIComponent(id)}/overview`);
export const createTeam = (body: unknown) =>
  api.post<{ team: Team }>("/api/v1/teams", body);
export const updateTeam = (id: string, body: unknown) =>
  api.patch<{ team: Team }>(`/api/v1/teams/${encodeURIComponent(id)}`, body);
export const addTeamMember = (
  teamId: string,
  body: {
    agentId: string;
    role: string;
    instructions?: string;
    runtimeBindingPolicy?: RuntimeBindingPolicy | null;
  },
) =>
  api.post<{ member: TeamMember }>(
    `/api/v1/teams/${encodeURIComponent(teamId)}/members`,
    body,
  );
export const updateTeamMember = (
  teamId: string,
  memberId: string,
  body: {
    role: string;
    instructions?: string;
    capabilityRequirements?: Record<string, unknown>;
    runtimeBindingPolicy?: RuntimeBindingPolicy | null;
    expectedTeamVersion: number;
  },
) => api.patch<{ member: TeamMember }>(
  `/api/v1/teams/${encodeURIComponent(teamId)}/members/${encodeURIComponent(memberId)}`,
  body,
);
export const removeTeamMember = (teamId: string, memberId: string) =>
  api.delete<void>(
    `/api/v1/teams/${encodeURIComponent(teamId)}/members/${encodeURIComponent(memberId)}`,
  );
export interface InboxSummary {
  unread: number;
  actionRequired: number;
  pendingApprovals: number;
  attentionTotal: number;
  byType: Record<string, number>;
}
export interface InboxOptions {
  view?: string;
  type?: string;
  archived?: string;
  cursor?: string;
  limit?: number;
}
export interface InboxPageResult {
  items: InboxItem[];
  hasMore: boolean;
  nextCursor: string;
}
export const listInbox = (tenant: string, namespace: string, options: InboxOptions = {}) =>
  api.get<InboxPageResult>(`/api/v1/inbox${query({ tenant, namespace, ...options })}`);
export const getInbox = (id: string, tenant: string, namespace: string) =>
  api.get<{ item: InboxItem }>(`/api/v1/inbox/${encodeURIComponent(id)}${query({ tenant, namespace })}`);
export const getInboxSummary = (tenant: string, namespace: string) =>
  api.get<{ summary: InboxSummary }>(`/api/v1/inbox/summary${query({ tenant, namespace })}`);
export const readInbox = (id: string, tenant?: string, namespace?: string) =>
  api.post<{ item: InboxItem }>(`/api/v1/inbox/${encodeURIComponent(id)}/read${query({ tenant, namespace })}`, {});
export const archiveInboxMessage = (id: string, tenant: string, namespace: string) =>
  api.post<{ item: InboxItem }>(`/api/v1/inbox/${encodeURIComponent(id)}/archive${query({ tenant, namespace })}`, {});
export const getApproval = (id: string) =>
  api.get<{ approval: Approval }>(`/api/v1/approvals/${encodeURIComponent(id)}`);
export const listApprovals = (
  tenant: string,
  namespace: string,
  status = "pending",
) =>
  api.get<{ items: Approval[] }>(
    `/api/v1/approvals${query({ tenant, namespace, status })}`,
  );
export const decideApproval = (
  id: string,
  status: string,
  expectedVersion: number,
  decision: unknown,
) =>
  api.post<{ approval: Approval }>(
    `/api/v1/approvals/${encodeURIComponent(id)}/decide`,
    { status, expectedVersion, decision },
  );
